File: src/main/java/nl/sijpesteijn/lasforce/services/AnimationService.java
package nl.sijpesteijn.lasforce.services;

import com.google.inject.Inject;
import com.google.inject.Provider;
import com.mongodb.BasicDBObject;
import com.mongodb.MongoClient;
import nl.sijpesteijn.ilda.Ilda;
import nl.sijpesteijn.ilda.IldaReader;
import nl.sijpesteijn.lasforce.domain.Animation;
import nl.sijpesteijn.lasforce.domain.LasForceAnimation;
import nl.sijpesteijn.lasforce.exceptions.LaserException;
import nl.sijpesteijn.lasforce.laser.IldaToLasForceConverter;
import org.apache.commons.configuration.Configuration;
import org.apache.commons.io.FileUtils;
import org.mongodb.morphia.Datastore;
import org.mongodb.morphia.Morphia;

import javax.inject.Singleton;
import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.Date;
import java.util.List;

/**
 * @author: Gijs Sijpesteijn
 */
@Singleton
public class AnimationService {
    private final String dataStore;
    private final int defaultFramerate;
    private IldaReader reader = new IldaReader();
    private Morphia morphia = new Morphia();
    private Datastore lasforce;

    @Inject
    public AnimationService(Provider<MongoClient> mongoProvider, Configuration configuration) {
        this.dataStore = configuration.getString("data_store");
        this.defaultFramerate = configuration.getInt("default_framerate");
        lasforce = morphia.createDatastore(mongoProvider.get(), "lasforce");
    }

    public Animation getAnimation(final long id) throws LaserException {
        Animation animation = lasforce.find(Animation.class).field("_id").equal(id).get();
        return animation;
    }

    public List<Animation> getAnimations() throws LaserException {
        List<Animation> animations = lasforce.find(Animation.class).asList();
        return animations;
    }


    public Animation addAnimation(Animation animation) throws LaserException {
        long id = new Date().getTime();
        BasicDBObject dbObject = new BasicDBObject("_id", id)
                .append("name", animation.getName())
                .append("lastUpdate", id)
                .append("frameRate", animation.getFrameRate())
                .append("fileName", animation.getFileName());

        lasforce.save(dbObject);
        animation.setId(id);
        animation.setLastUpdate(id);
        return animation;
    }

    public void update(final Animation animation) throws LaserException {
    }

    public Animation uploadAnimation(String fileName, byte[] data) throws LaserException {
        Animation animation = new Animation();
        try {
            FileUtils.writeByteArrayToFile(new File(this.dataStore + fileName), data);
            animation.setName(fileName);
            animation.setFrameRate(defaultFramerate);
            animation.setFileName(fileName);
            animation = addAnimation(animation);
            return animation;
        } catch (IOException e) {
            throw new LaserException(e);
        }
    }

    public LasForceAnimation getLasForceAnimation(long id) throws LaserException {
        Animation animation = getAnimation(id);
        try {
            Ilda ilda = reader.read(new File(this.dataStore + animation.getFileName()));
            IldaToLasForceConverter converter = new IldaToLasForceConverter();
            AnimationMetaData metaData = new AnimationMetaData();
            metaData.setFrameRate(animation.getFrameRate());
            metaData.setName(animation.getName());
            LasForceAnimation lasForceAnimation = converter.convert(ilda);
            lasForceAnimation.setMetaData(metaData);
            return lasForceAnimation;
        } catch (IOException e) {
            throw new LaserException(e);
        } catch (URISyntaxException e) {
            throw new LaserException(e);
        }
    }

}

File: src/main/java/nl/sijpesteijn/lasforce/services/AnimationService.kt
package nl.sijpesteijn.lasforce.services

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.mongodb.BasicDBObject
import com.mongodb.DBCollection
import com.mongodb.DBObject
import nl.sijpesteijn.ilda.IldaReader
import nl.sijpesteijn.lasforce.IldaConverter
import nl.sijpesteijn.lasforce.domain.Animation
import nl.sijpesteijn.lasforce.guice.MongoDBProvider
import org.apache.commons.configuration.Configuration
import org.apache.commons.io.FileUtils
import java.io.File
import java.util.*
import javax.inject.Inject

/**
 * @author Gijs Sijpesteijn
 */
class AnimationService {
    private val collection: DBCollection
    private val dataStore: String
    private val defaultFramerate: Int
    private val mapper = jacksonObjectMapper()

    @Inject
    constructor(mongoDBProvider: MongoDBProvider, configuration: Configuration) {
        this.collection = mongoDBProvider.getDB().getCollection("animations")
        this.dataStore = configuration.getString("data_store")
        this.defaultFramerate = configuration.getInt("default_framerate")
    }

    fun getAnimationsMetadata(): List<Animation> {
        val animations = ArrayList<Animation>()
        val cursor = collection.find()
        while (cursor.hasNext()) {
            val dbObject = cursor.next()
            val animation = loadMetadata(dbObject)
            animations.add(animation)
        }
        return animations
    }

    fun getAnimationMetadata(id: Long): Animation {
        val dbObject = collection.findOne(BasicDBObject("_id", id))
        return loadMetadata(dbObject)
    }

    private fun loadMetadata(dbObject:DBObject) : Animation {
        val id = dbObject.get("_id") as Long
        val name = dbObject.get("name") as String
        val last_update = dbObject.get("last_update") as Long
        val repeat = dbObject.get("repeat") as Int
        val total_frames = dbObject.get("total_frames") as Int
        val animation = Animation(id, name, last_update, repeat, total_frames)
        return animation
    }

    fun getLasForceAnimation(id: Long): Animation {
        val animation: Animation = mapper.readValue(FileUtils.readFileToString(File(this.dataStore + id + ".las")))
        return animation
    }

    fun uploadAnimation(fileName: String, data: ByteArray?) {
        val name = fileName.split(".")[0];
        val convert = IldaConverter()
        val reader = IldaReader()
        val ilda = reader.read(data)
        val frames = convert.fromIlda(ilda)
        val animation = Animation(System.currentTimeMillis(), name, System.currentTimeMillis(), 1, frames.size, 1000, frames)
        saveLasforceAnimation(animation)
    }

    fun saveLasforceAnimation(animation: Animation): Animation {
        val new_animation = saveAnimationMetadata(animation)
        mapper.writeValue(File(this.dataStore + new_animation.id + ".las"), animation)
        return new_animation
    }

    fun saveAnimationMetadata(animationMetadata: Animation): Animation {
        animationMetadata.last_update = System.currentTimeMillis()
        val dbObject = BasicDBObject("_id", animationMetadata.id)
                .append("name", animationMetadata.name)
                .append("last_update", animationMetadata.last_update)
                .append("total_frames", animationMetadata.total_frames)
                .append("repeat", animationMetadata.repeat)
        collection.save(dbObject)
        return animationMetadata
    }

    fun deleteAnimation(id: Long) {
        File(this.dataStore + "${id}.las").delete();
        collection.remove(BasicDBObject("_id", id));
    }
}